File: csharp/TellDontAskKata.Domain/Order.cs
﻿using System;
using System.Collections.Generic;

namespace TellDontAskKata.Domain
{
    public class Order
    {
        private decimal total;
        private String currency;
        private List<OrderItem> items;
        private decimal tax;
        private OrderStatus status;
        private int id;

        public Order()
        {
            SetStatus(OrderStatus.Created);
            SetItems(new List<OrderItem>());
            SetTotal((decimal)0.0);
            SetCurrency("EUR");
            SetTax((decimal)0.0);
        }

        public void AddItem(OrderItem item)
        {
            GetItems().Add(item);

            if (HasTooManyFoodItems())
            {
                throw new MaximumNumberOfFoodItemsExceeded();
            }

            SetTotal(GetTotal() + item.GetTaxedAmount());
            SetTax(GetTax() + item.GetTax());
        }

        public bool HasTooManyFoodItems()
        {
            int numberOfFoodItems = 0;
            foreach (OrderItem item in GetItems())
            {
                if (item.GetProduct().GetCategory().GetName().Equals("food"))
                {
                    numberOfFoodItems += item.getQuantity();
                }
            }

            return numberOfFoodItems > 100;
        }

        public decimal GetTotal()
        {
            return total;
        }

        public void SetTotal(decimal total)
        {
            this.total = total;
        }

        public String GetCurrency()
        {
            return currency;
        }

        public void SetCurrency(String currency)
        {
            this.currency = currency;
        }

        public List<OrderItem> GetItems()
        {
            return items;
        }

        public void SetItems(List<OrderItem> items)
        {
            this.items = items;
        }

        public decimal GetTax()
        {
            return tax;
        }

        public void SetTax(decimal tax)
        {
            this.tax = tax;
        }

        public OrderStatus GetStatus()
        {
            return status;
        }

        public void SetStatus(OrderStatus status)
        {
            this.status = status;
        }

        public int GetId()
        {
            return id;
        }

        public void SetId(int id)
        {
            this.id = id;
        }
    }
}

File: java/src/main/java/it/gabrieletondi/telldontaskkata/domain/Order.java
package it.gabrieletondi.telldontaskkata.domain;

import static it.gabrieletondi.telldontaskkata.domain.OrderStatus.CREATED;
import static it.gabrieletondi.telldontaskkata.domain.OrderStatus.REJECTED;
import static it.gabrieletondi.telldontaskkata.domain.OrderStatus.SHIPPED;

import it.gabrieletondi.telldontaskkata.service.ShipmentService;
import it.gabrieletondi.telldontaskkata.useCase.*;

import java.math.BigDecimal;
import java.util.List;

public class Order {
	private BigDecimal total;
	private String currency;
	private List<OrderItem> items;
	private BigDecimal tax;
	private OrderStatus status;
	private int id;

	public BigDecimal getTotal() {
		return total;
	}

	public void setTotal(BigDecimal total) {
		this.total = total;
	}

	public String getCurrency() {
		return currency;
	}

	public void setCurrency(String currency) {
		this.currency = currency;
	}

	public List<OrderItem> getItems() {
		return items;
	}

	public void setItems(List<OrderItem> items) {
		this.items = items;
	}

	public BigDecimal getTax() {
		return tax;
	}

	public void setTax(BigDecimal tax) {
		this.tax = tax;
	}

	public OrderStatus getStatus() {
		return status;
	}

	public void setStatus(OrderStatus status) {
		this.status = status;
	}

	public int getId() {
		return id;
	}

	public void setId(int id) {
		this.id = id;
	}

	private void verifyShippable() {
		if (getStatus().equals(CREATED) || getStatus().equals(REJECTED)) {
			throw new OrderCannotBeShippedException();
		}

		if (getStatus().equals(SHIPPED)) {
			throw new OrderCannotBeShippedTwiceException();
		}
	}

	public void approve() {
		verifyCanBeChanged();
		if (getStatus().equals(REJECTED)) {
			throw new RejectedOrderCannotBeApprovedException();
		}

		setStatus(OrderStatus.APPROVED);
	}

	public void reject() {
		verifyCanBeChanged();
		if (getStatus().equals(OrderStatus.APPROVED)) {
			throw new ApprovedOrderCannotBeRejectedException();
		}

		setStatus(OrderStatus.REJECTED);
	}

	private void verifyCanBeChanged() {
		if (getStatus().equals(SHIPPED)) {
			throw new ShippedOrdersCannotBeChangedException();
		}
	}

	public void ship(ShipmentService shipmentService) {
		verifyShippable();

		shipmentService.ship(this);

		setStatus(SHIPPED);
	}
}
